metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala

```diff
@@ -0,0 +1,206 @@
+package scala.meta.metals.lsp
+
+import java.util
+import java.util.concurrent.CompletableFuture
+
+import scala.meta.internal.metals.DidFocusResult
+import scala.meta.internal.metals.HoverExtParams
+import scala.meta.internal.metals.WindowStateDidChangeParams
+import scala.meta.internal.metals.doctor.DoctorVisibilityDidChangeParams
+import scala.meta.internal.metals.findfiles.FindTextInDependencyJarsRequest
+import scala.meta.internal.tvp.MetalsTreeViewChildrenResult
+import scala.meta.internal.tvp.TreeViewChildrenParams
+import scala.meta.internal.tvp.TreeViewNodeCollapseDidChangeParams
+import scala.meta.internal.tvp.TreeViewNodeRevealResult
+import scala.meta.internal.tvp.TreeViewParentParams
+import scala.meta.internal.tvp.TreeViewParentResult
+import scala.meta.internal.tvp.TreeViewVisibilityDidChangeParams
+
+import org.eclipse.lsp4j._
+import org.eclipse.lsp4j.jsonrpc.messages.{Either => JEither}
+
+/**
+ * Delegating Scala LSP service which forwards all requests to the underlying
+ * instance. This is needed to support lsp4j
+ * [[org.eclipse.lsp4j.jsonrpc.services.JsonDelegate]]. Value returned by
+ * method with that annotation is used to determine which methods are supported
+ * by the server. We can't initialize it to null because then no lsp methods would be
+ * supported. Instead, we initialize it to a dummy instance which is then replaced.
```

Comment:
Bear with me, I'll probably ask a bunch of these just t make sure I'm tracking, but in case the dummy on is the `new ScalaLspService {}` which will then later be replaced by the `MetalsLspService` as the _real_ one right?

Reply:
yes, you're correct.
The thing is, at runtime (I'm not sure at exactly which point though), probably betwenn those lines https://github.com/kpodsiad/metals/blob/908c0c1e199626abcf3375d96555772568e794ab/metals/src/main/scala/scala/meta/metals/Main.scala#L59-L70 lsp4j processes given server which was set in `setLocalService` method. If method with `JsonDelegate` annotation exists, that method is being called. If it returns null processing is aborted, if not them processing is continued on class which is being returned. 

The tricky part - now we delay creation of MetalsLspService so we cannot provide working instance from the very beginning. We can't return null because lsp4j processing won't process `ScalaLspService` methods and our MetalsLanguageServer will "see" only basic init/shutdown endpoints*

* - it is also possible that I misunderstood some parts of how lsp4j works, but after I reached current state I was happy enough to see things working.

Answering your questions in this thread so we can keep discussion in one place:

`ScalaLspService` isn't just a Dummy instance, it's an interface which describes endpoints. That interface is being later on implemented by MetalsLspService. What may be con fusing is fact, that `ScalaLspService` mixes two roles:
- it's interface
- it provides default implementation for methods so it is possible to do just `new ScalaLspService {}`

Default implementations are needed because of later point. Otherwise all of those methods had to be implemented

Are things more clear now? Maybe scaladocs needs some improvement?

Reply:
Yea this helps, thanks.

Reply:
Should I modify scaladoc to make things crystal clear?

Reply:
I think more than a Scaladoc we should try and outline the entire thing in the design doc.